python/cuml/experimental/explainer/base.pyx

```diff
@@ -0,0 +1,406 @@
+#
+# Copyright (c) 2020-2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cudf
+import cuml.internals
+import cupy as cp
+import numpy as np
+import pandas
+
+import cuml.common.logger as logger
+from cuml.common.import_utils import has_shap
+from cuml.common.input_utils import input_to_cupy_array
+from cuml.common.logger import warn
+from cuml.experimental.explainer.common import get_dtype_from_model_func
+from cuml.experimental.explainer.common import get_handle_from_cuml_model_func
+from cuml.experimental.explainer.common import get_link_fn_from_str_or_fn
+from cuml.experimental.explainer.common import get_tag_from_model_func
+from cuml.experimental.explainer.common import model_func_call
+from cuml.experimental.explainer.common import output_list_shap_values
+
+from cuml.raft.common.handle cimport handle_t
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+
+
+cdef extern from "cuml/explainer/permutation_shap.hpp" namespace "ML":
+
+    void shap_main_effect_dataset "ML::Explainer::shap_main_effect_dataset"(
+        const handle_t& handle,
+        float* dataset,
```

Comment:
(in the underlying API) should dataset be const?

Reply:
dataset is where the output is generated, maybe I should change the name to avoid the confusion?